pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/redirects.py

```diff
@@ -0,0 +1,238 @@
+                        client_side_redirects[location] = f"{xref_targets[identifier].path}#{identifier}"
+                    else:
+                        divergent_redirects.add(location)
+                else:
+                    if location not in server_side_redirects:
+                        server_side_redirects[location] = xref_targets[identifier].path
+                    else:
+                        divergent_redirects.add(location)
+        if divergent_redirects:
+            self._errors.append(DivergentRedirectError(divergent_redirects))
+        if identifiers_missing_current_outpath:
+            self._errors.append(InvalidCurrentPathError(identifiers_missing_current_outpath))
+        if identifiers_without_redirects:
+            self._errors.append(MissingRedirectError(identifiers_without_redirects))
+
+        for target in {location.split('#')[0] for location in all_client_locations}:
+            identifiers = [identifier for identifier, xref in xref_targets.items() if xref.path == target]
+            anchors = {location.split('#')[1] for location in all_client_locations}
+            conflicting_anchors |= anchors.intersection(identifiers)
+        if conflicting_anchors:
+            self._errors.append(ConflictingAnchorsError(conflicting_anchors))
+
+        transitive_redirects = {}
+        for server_from, server_to in server_side_redirects.items():
+            for client_from, client_to in client_side_redirects.items():
+                path, anchor = client_from.split('#')
+                if server_from == path:
+                    transitive_redirects[client_from] = f"{server_to}#{anchor}"
+        if transitive_redirects:
+            self._errors.append(TransitiveRedirectError(transitive_redirects))
+
+    def raise_errors(self, error_type: RedirectsError = None):
+        """
+        Raises all errors if error_type is not specified.
+        If error_type is specified, only raises error of that type.
+        """
+        errors_to_raise = [error for error in self._errors if error_type is None or isinstance(error, error_type)]
+        if len(errors_to_raise) == 0:
+            return
+        if len(errors_to_raise) == 1:
+            raise errors_to_raise[0]
+        else:
+            raise RedirectsError(f"Multiple validation errors occurred:\n{'\n'.join(map(str, errors_to_raise))}")
+
+    def get_client_redirects(self, redirection_target: str):
+        client_redirects = {}
+        for identifier, locations in self._raw_redirects.items():
+            for location in locations[1:]:
+                if '#' not in location:
+                    continue
+                path, anchor = location.split('#')
+                if path != redirection_target:
+                    continue
+                client_redirects[anchor] = f"{locations[0]}#{identifier}"
+
+        return self._redirects_script.replace('REDIRECTS_PLACEHOLDER', json.dumps(client_redirects))
```

Comment:
```suggestion
        return self._redirects_script.replace('REDIRECTS_PLACEHOLDER', json.dumps(client_redirects), sort_keys=True)
```

We should keep the order predictable, otherwise we'll surely run into issues later on. Right now it doesn't show because we're reading in file order and writing in processing order, but who knows how this code will evolve.